gwpy/timeseries/timeseries.py

```diff
@@ -1567,6 +1606,98 @@ def whiten(self, fftlength=None, overlap=0, method=DEFAULT_FFT_METHOD,
         out = in_.convolve(tdw, window=window)
         return out * numpy.sqrt(2 * in_.dt.decompose().value)
 
+    def gate(self, tzero=1.0, tpad=0.5, whiten=True,
+             threshold=50., cluster_window=0.5, **whiten_kwargs):
+        """Discovers a set of indices where gating windows should be placed
+           based on a given amplitude threshold. Points are clustered within
+           a time window.
+
+        Parameters
+        ----------
+        tzero : `int`, optional
+            half-width time duration in which the time series is set to zero
+
+        tpad : `int`, optional
+            half-width time duration in which the Planck window is tapered
+
+        whiten : `bool`, optional
+            if True, data will be whitened before gating points are discovered,
+            use of this option is highly recommended
+
+        threshold : `float`, optional
+            amplitude threshold, if the data exceeds this value a gating window
+            will be placed
+
+        cluster_window : `float`, optional
+            time duration over which gating points will be clustered
+
+        **whiten_kwargs
+            other keyword arguments that will be passed to the
+            `TimeSeries.whiten` method if it is being used when discovering
+            gating points
+
+        Returns
+        -------
+        out : `~gwpy.timeseries.TimeSeries`
+            a copy of the original `TimeSeries` that has had gating windows
+            applied
+
+        Examples
+        --------
+
+        Read data into a `TimeSeries`
+        >>> from gwpy.timeseries import TimeSeries
+        >>> data = TimeSeries.get('H1:GDS-CALIB_STRAIN', 1225542623,
+                                   1225542823)
+
+        Apply gating using custom arguments
+        >>> gated = data.gate(tzero=1.0, tpad=1.0, threshold=10.0,
+                              fftlength=4, overlap=2, method='median')
+
+        Plot the original data and the gated data
+        >>> overlay = data.plot(dpi=150, label='Ungated',
+                                color='dodgerblue', zorder=3)
+        >>> ax = overlay.gca()
+        >>> ax.plot(gated, label='Gated', color='orange', zorder=2)
+        >>> ax.set_xlim(1225542713, 1225542733)
+        >>> ax.legend()
+        >>> overlay.show()
+        """
+        # Find points to gate based on a threshold
+        if whiten:
+            gates = _discover_gate_points(self.whiten(**whiten_kwargs),
+                                          threshold, cluster_window)
+        else:
+            gates = _discover_gate_points(self, threshold, cluster_window)
+        out = self.copy()
+
+        # Iterate over list of indices to gate and apply each one
+        for gate in gates:
+            nzero = int(abs(tzero)*self.sample_rate.value)
+            npad = int(abs(tpad)*self.sample_rate.value)
+            half = nzero + npad
+            ntotal = 2*half
```

Comment:
it looks like all of this is gate-independent, so could be moved outside the loop.